src/shogun/base/base_types.h

```diff
@@ -68,6 +70,121 @@ namespace shogun
 	                  const char*, typename std::decay<T>::type>::value>
 	{
 	};
+
+	// Primitive types
+	enum class TYPE
+	{
+		T_BOOL = 1,
+		T_CHAR = 2,
+		T_INT8 = 3,
+		T_UINT8 = 4,
+		T_INT16 = 5,
+		T_UINT16 = 6,
+		T_INT32 = 7,
+		T_UINT32 = 8,
+		T_INT64 = 9,
+		T_UINT64 = 10,
+		T_FLOAT32 = 11,
+		T_FLOAT64 = 12,
+		T_FLOATMAX = 13,
+		T_SGOBJECT = 14,
+		T_COMPLEX128 = 15,
+		T_SGVECTOR_FLOAT32 = 16,
+		T_SGVECTOR_FLOAT64 = 17,
+		T_SGVECTOR_FLOATMAX = 18,
+		T_SGVECTOR_INT32 = 19,
+		T_SGVECTOR_INT64 = 20,
+		T_SGMATRIX_FLOAT32 = 21,
+		T_SGMATRIX_FLOAT64 = 22,
+		T_SGMATRIX_FLOATMAX = 23,
+		T_SGMATRIX_INT32 = 24,
+		T_SGMATRIX_INT64 = 25,
+		T_UNDEFINED = 26
+	};
+
+
+	template <class T> class SGVector;
+	template <class T> class SGMatrix;
+
+	namespace type_internal
+	{
+		template <typename T>
+		struct sg_type
+		{
+		};
+
+		template <typename T>
+		struct is_sg_primitive : public std::false_type
+		{
+		};
+
+		template <typename T>
+		struct is_sg_vector : public std::false_type
+		{
+		};
+
+		template <typename T>
+		struct is_sg_matrix : public std::false_type
+		{
+		};
```

Comment:
Remember that these structs do not check all possible `T`/`SGVector<T>`/`SGMatrix<T>`, this is to check if the type is registered at runtime to then use the dispatching. To check if a type is `SGVector<T>` you should do something like:
```
template <typename T>
struct is_sg_vector
{
    using type = std::false_type;
};
template <typename T>
struct is_sg_vector<SGVector<T>>
{
  using type = std::true_type;
};
```

Reply:
(That's why these things were hidden in their own namespace)

Reply:
Argh you are indeed right. Those need to be left in their namespace. In the end, I just need to check if it is an SGVector/SGMatrix (an not if they match also with a type `T`).